File: NeetCode 150/Graphs/332. Reconstruct Itinerary.py
class Solution:
    def findItinerary(self, tickets: List[List[str]]) -> List[str]:
        # Graph where:
        # key   = source airport
        # value = min-heap of destination airports
        # Min-heap ensures lexicographically smallest destination is chosen first
        graph = defaultdict(list)

        for src, dst in tickets:
            heapq.heappush(graph[src],dst)
        # This will store the itinerary in reverse order
        route = []

        def dfs(airport):
            while graph[airport]:
                """
            Hierholzer's DFS:
            - Keep visiting next smallest destination
            - Remove (use) each ticket exactly once
            - Add airport to route ONLY after all outgoing edges are used
            """
                # Choose the lexicographically smallest destination
                new_airport = heapq.heappop(graph[airport])
                dfs(new_airport)
            # Postorder insertion:
            # When no outgoing edges remain, add airport to route
            route.append(airport)
        
        dfs("JFK")
        return route[::-1]  # Reverse because we added airports after exhausting all flights

File: NeedCode 150 (JAVA)/Graphs/332. Reconstruct Itinerary.java
class Solution {
    public List<String> findItinerary(List<List<String>> tickets) {
        Map<String, PriorityQueue<String>> graph = new HashMap<>();

        for(List<String> ticket: tickets){
            String from  = ticket.get(0);
            String to = ticket.get(1);
            graph.putIfAbsent(from,new PriorityQueue<>());
            graph.get(from).offer(to);
        }

        List<String> result = new ArrayList<>();
        dfs("JFK",graph,result);
        Collections.reverse(result);
        return result;
    }
    private void dfs(String airport, Map<String,PriorityQueue<String>> graph, List<String> result){
        PriorityQueue<String> pq = graph.get(airport);

        while(pq!=null && !pq.isEmpty()){
            String next = pq.poll();
            dfs(next,graph,result);
        }
        result.add(airport);
    }
}
